src/java.base/share/classes/java/text/MessageFormat.java

```diff
@@ -1651,6 +1819,90 @@ private static void copyAndFixQuotes(String source, int start, int end,
         }
     }
 
+    // Corresponding to the FormatType pattern
+    private enum FormatType {
+        NUMBER("number"),
+        DATE("date"),
+        TIME("time"),
+        DTF_DATE("dtf_date"),
+        DTF_TIME("dtf_time"),
+        DTF_DATETIME("dtf_datetime"),
+        CHOICE("choice"),
+        LIST("list"),
+
+        // Pre-defined DateTimeFormatter types
+        BASIC_ISO_DATE("BASIC_ISO_DATE"),
+        ISO_LOCAL_DATE("ISO_LOCAL_DATE"),
+        ISO_OFFSET_DATE ("ISO_OFFSET_DATE"),
+        ISO_DATE("ISO_DATE"),
+        ISO_LOCAL_TIME("ISO_LOCAL_TIME"),
+        ISO_OFFSET_TIME("ISO_OFFSET_TIME"),
+        ISO_TIME("ISO_TIME"),
+        ISO_LOCAL_DATE_TIME("ISO_LOCAL_DATE_TIME"),
+        ISO_OFFSET_DATE_TIME("ISO_OFFSET_DATE_TIME"),
+        ISO_ZONED_DATE_TIME("ISO_ZONED_DATE_TIME"),
+        ISO_DATE_TIME("ISO_DATE_TIME"),
+        ISO_ORDINAL_DATE("ISO_ORDINAL_DATE"),
+        ISO_WEEK_DATE("ISO_WEEK_DATE"),
+        ISO_INSTANT("ISO_INSTANT"),
+        RFC_1123_DATE_TIME("RFC_1123_DATE_TIME");
+
+        private final String text;
+
+        FormatType(String text){
+            this.text = text;
+        }
+
+        // This method returns a FormatType that matches the passed String.
+        // If no matching FormatType is found, an IllegalArgumentException is thrown.
+        private static FormatType fromString(String text) {
+            for (FormatType other : values()) {
+                // Also check trimmed lower case for historical reasons
+                if (text.trim().toLowerCase(Locale.ROOT)
+                        .equals(other.text.trim().toLowerCase(Locale.ROOT))) {
+                    return other;
+                }
+            }
+            throw new IllegalArgumentException();
+        }
+    }
+
+    // Corresponding to the FormatStyle pattern
+    private enum FormatStyle {
+        DEFAULT(""),
+        SHORT("short"),
+        MEDIUM("medium"),
+        LONG("long"),
+        FULL("full"),
+        INTEGER("integer"),
+        CURRENCY("currency"),
+        PERCENT("percent"),
+        COMPACT_SHORT("compact_short"),
+        COMPACT_LONG("compact_long"),
+        OR("or"),
+        UNIT("unit"),
+        SUBFORMATPATTERN(null);
+
+        private final String text;
+
+        FormatStyle(String text){
+            this.text = text;
+        }
+
+        // This method returns a FormatStyle that matches the passed String.
+        // If no FormatStyle is found, IllegalArgumentException is thrown
+        private static FormatStyle fromString(String text) {
+            for (FormatStyle style : values()) {
+                // Also check trimmed lower case for historical reasons
+                if (text.equals(style.text)
+                        || text.trim().toLowerCase(Locale.ROOT).equals(style.text)) {
+                    return style;
+                }
+            }
+            throw new IllegalArgumentException();
+        }
+    }
```

Comment:
Same as above.

Reply:
I would rather use `fromString(String text)` here, as using `Enum.valueOf()` does not work as smoothly as `FormatType`, since the `FormatStyle.text` value does not always match the `FormatStyle` enum name. 

For example, `FormatStyle.DEFAULT` has a `.text` value of "". So I would have to convert a "" to "default" to check for `FormatStyle.DEFAULT` via `Enum.valueOf()`, but at the same time, this also means that the behavior would now change to accept "default" as `FormatStyle.DEFAULT` (which is not in the spec).

I could then add extra code to handle the "default" case, but I think at that point it would be much too much complexity.

Reply:
Let's keep this static method then.